jvb/src/main/kotlin/org/jitsi/videobridge/xmpp/ClientConnection.kt

```diff
@@ -0,0 +1,199 @@
+            mucClientManager.addMucClient(config)
+        }
+
+        // We consider the case where a client with the given ID already
+        // exists as success. Note however, that the existing client's
+        // configuration was NOT modified.
+        return true
+    }
+
+    fun removeMucClient(jsonObject: JSONObject): Boolean {
+        if (jsonObject["id"] !is String) {
+            return false
+        }
+        return mucClientManager.removeMucClient(jsonObject["id"] as String)
+    }
+
+    /**
+     * Process an incoming IQ
+     */
+    override fun handleIq(iq: IQ?): IQ? {
+        if (iq == null) {
+            return null
+        }
+        logger.cdebug { "RECV: ${iq.toXML()}" }
+
+        val response = when (iq.type) {
+            IQ.Type.get, IQ.Type.set -> handleIqRequest(iq)
+            else -> null
+        }
+
+        logger.cdebug { "SENT: ${response?.toXML() ?: "null"}" }
+
+        return response
+    }
+
+    private fun handleIqRequest(iq: IQ): IQ {
+        val handler = eventHandler ?: return IQUtils.createError(
+            iq,
+            XMPPError.Condition.service_unavailable,
+            "Service unavailable"
+        )
+        return when (iq) {
+            is Version -> measureDelay(versionDelayStats, { iq.toXML() }) {
+                handler.versionIqReceived(iq)
+            }
+            is ColibriConferenceIQ -> measureDelay(colibriDelayStats, { iq.toXML() }) {
+                handler.colibriConferenceIqReceived(iq)
+            }
+            is HealthCheckIQ -> measureDelay(healthDelayStats, { iq.toXML() }) {
+                handler.healthCheckIqReceived(iq)
+            }
+            else -> IQUtils.createError(
+                iq,
+                XMPPError.Condition.service_unavailable,
+                "Unsupported IQ request ${iq.childElementName}"
+            )
+        }.apply {
+            from = iq.to
+            stanzaId = iq.stanzaId
+            to = iq.from
+        }
+    }
+
+    private fun <T> measureDelay(delayStats: DelayStats, context: () -> CharSequence, block: () -> T): T {
+        val start = System.currentTimeMillis()
+        val result = block()
+        val delayMs = System.currentTimeMillis() - start
+        delayStats.addDelay(delayMs)
+        if (delayMs > 100) {
+            logger.warn("Took $delayMs ms to handle IQ: ${context()}")
+        }
+        return result
+    }
+
+    interface EventHandler {
+        fun colibriConferenceIqReceived(iq: ColibriConferenceIQ): IQ
+        fun versionIqReceived(iq: Version): IQ
+        fun healthCheckIqReceived(iq: HealthCheckIQ): IQ
+    }
+
+    companion object {
+        private val FEATURES = arrayOf(
+            ColibriConferenceIQ.NAMESPACE,
+            HealthCheckIQ.NAMESPACE,
+            "urn:xmpp:jingle:apps:dtls:0",
+            "urn:xmpp:jingle:transports:ice-udp:1",
+            Version.NAMESPACE
+        )
+
+        private val delayThresholds = longArrayOf(5, 50, 100, 1000)
```

Comment:
Note: these were static to accomodate both XMPP client and component modes. They don't need to be static anymore (but that doesn't need to change in this PR).

Reply:
Ok, yeah I think I'll leave those as-is for now.